src/util/server.ml

```diff
@@ -0,0 +1,184 @@
+open Batteries
+open Jsonrpc
+
+exception Failure of Response.Error.Code.t * string
+
+type t = {
+  file: Cil.file;
+  do_analyze: Analyses.increment_data -> Cil.file -> unit;
+  input: IO.input;
+  output: unit IO.output;
+}
+
+module type Request = sig
+  val name: string
+
+  type params
+  type response
+
+  val params_of_yojson: Yojson.Safe.t -> (params, string) result
+  val response_to_yojson: response -> Yojson.Safe.t
+
+  val process: params -> t -> response
+end
+
+module Registry = struct
+  type t = (string, (module Request)) Hashtbl.t
+  let make () : t = Hashtbl.create 32
+  let register (reg: t) (module R : Request) = Hashtbl.add reg R.name (module R)
+end
+
+let registry = Registry.make ()
+
+let handle_exn id exn =
+  Response.Error.(make Code.InternalError (Printexc.to_string exn) () |> Response.error id)
+
+module ParamParser (R : Request) = struct
+  let parse params =
+    let maybe_params =
+      params
+      |> Option.map_default Message.Structured.to_json `Null
+      |> R.params_of_yojson
+    in
+    match maybe_params with
+    | Ok params -> Ok params
+    | Error err ->
+      (* This is a hack to handle cases where R.params is a primitive type like int or string. *)
+      match params with
+      | Some `List [param] -> R.params_of_yojson param |> Result.map_error (fun _ -> err)
+      | _ -> Error err
+end
+
+let handle_request (serv: t) (message: Message.either) (id: Id.t) =
+  let req = Hashtbl.find_option registry message.method_ in
+  let response = match req with
+    | Some (module R) ->
+      let module Parser = ParamParser (R) in (
+        match Parser.parse message.params with
+        | Ok params -> (
+            try
+              R.process params serv
+              |> R.response_to_yojson
+              |> Response.ok id
+            with Failure (code, msg) -> Response.Error.(make code msg () |> Response.error id))
+        | Error s -> Response.Error.(make Code.InvalidParams s () |> Response.error id))
+    | _ -> Response.Error.(make Code.MethodNotFound message.method_ () |> Response.error id)
+  in
+  Response.yojson_of_t response |> Yojson.Safe.to_string |> IO.write_line serv.output;
+  IO.flush serv.output
+
+let serve serv =
+  serv.input
+  |> IO.to_input_channel
```

Comment:
@keremc  The documentation on `BatIO.to_input_channel` says that the operation is "extremely costly and is provided essentially for debugging purposes", see [here](https://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatIO.html). As `Yojson.Safe.linestream_from_channel` requires an `Stdlib.in_channel` maybe one just change `serv.input` to be of type `in_channel`?

Reply:
If anyone's wondering how costly is _extremely_, then here's the answer: https://github.com/ocaml-batteries-team/batteries-included/blob/a7276a8d41b31035423235b2395aa82ab5d6f392/src/batIO.ml#L687-L716 — it might go through a temporary file...

Reply:
Oops. Thanks for pointing this out. I plan to make a server mode-related PR soon anyway, so I'll include the fix in it.